### DISCUSSION TITLE
Experiment: replacing modernc.org/sqlite with github.com/ncruces/go-sqlite3 in gotosocial

### DISCUSSION THREAD

**Discussion by NyaaaWhatsUpDoc** [ncruces/go-sqlite3#78] [Show and tell]
Just starting an open discussion for this so I can document things as I experience them, and serves as a useful space for any related discussion to this Go sqlite implementation. I'm not 100% sure this is something we'd look at PR'ing to main, but there's definitely a very spirited discussion between our core devs on matrix :D.

So far there's a working build replacing modernc.org/sqlite with github.com/ncruces/go-sqlite3 here: https://github.com/NyaaaWhatsUpDoc/gotosocial/tree/experiment/use-wasm-sqlite3

I might look at upstreaming the funky little driver wrapper I made (see `./internal/db/sqlite`) as it makes wrapping the driver yourself easier if you need to, and i've kept our standard `retryOnBusy()` handler in place over using the built-in busy-timeout handler both due to it not being context aware (though change should be easy enough, i have a change for y'all ready if you want it upstream) and as i'm not sure if they 100% replace one-another.

This working build runs gotosocial as expected, with decent performance, minus a memory leak (which *doesn't* occur on current main) which causes the application to get oom killed every few hours. With the fact we're using `github.com/uptrace/bun` (which in turn wraps `database/sql`, which then uses our busy / error handling wrapper...) it's hard to narrow down exactly where it's being introduced, but i shall keep looking!

Anyways, i hope at least someone here in this discussion board finds this interesting! Either way i'll probably continue to ramble with my progress as i make it :)

**Comment by ncruces** [comment]
Not sure this is the best forum for this discussion (rather than something under [`github.com/superseriousbusiness/gotosocial`](https://github.com/superseriousbusiness/gotosocial)), where it might have more visibility of interested parties? I don't mind hosting it here either, so that's fine.

Any specific questions/concerns you might have for me, please make them more explicit, 'cause at first glance I don't think I can contribute much to the discussion.

I'll still open a few threads for things that come up.

**Reply by NyaaaWhatsUpDoc** [reply]
I'm one of the main gotosocial developers, just we don't really use our discussion board and I figured this project's discussion board might be more useful for others looking to make the switch and seeing this thread, whether it provides info or just encouragement. Though ultimatley I'm not hugely fussed where it ends up so I'm happy if you'd rather move it :)

**Comment by ncruces** [comment]
### Busy handler not being context aware

The philosophy of this driver is that the busy handler is the **maximum** you'll _ever_ want to wait. Queries should still bail out early if a context is cancelled. If this doesn't happen, I consider it a bug.

Acting on context done isn't immediate (it's hard to interrupt file IO, and I don't even try). Also, the default timeout handler sleeps increments of 100ms, so you can always miss cancellation by at least that much. But I don't think you should ever miss context cancellation by more than (e.g.) 250ms.

**Reply by NyaaaWhatsUpDoc** [reply]
Does the busy handler with a timeout set handle retries? Having looked at the sqlite docs multiple times for this I'm not 100% sure whether it does or not. If it does then it seems with it enabled it should render our `retryOnBusy()` wrapping code useless, which would be great since less wrapping code :p

**Reply by ncruces** [reply]
Yes, that's the entire point of it.

If the database is locked (returning [`SQLITE_BUSY`](https://www.sqlite.org/rescode.html#busy), [`SQLITE_LOCKED`](https://www.sqlite.org/rescode.html#locked) mostly applies to shared cache which we don't use), the _default_ busy handler will sleep in increments of at most 100ms, and retry up to the busy timeout. If you set the busy timeout high enough (I recommend 1min), you can then use contexts to cancel early instead (the error in this case will be [`SQLITE_INTERRUPT`](https://www.sqlite.org/rescode.html#interrupt)).

The only situations where this doesn't happen are those that would lead to [dead lock](https://www.sqlite.org/c3ref/busy_handler.html): two connections waiting "forever" on each other until one gives up. Then, one of the connections returns `SQLITE_BUSY` immediately so the other can proceed.

If this is returned by our driver _before_ the timeout expires, you're in the middle of a transaction where you read data that might be/become stale, and you need to _abort the entire transaction_ to retry. This is not something that can be done transparently.

However, the above situation can be avoided in all cases.

To avoid `SQLITE_BUSY` ever being returned, you need to follow one simple rule: don't upgrade read transactions into write transactions. If one transaction is going to write to any attached database, use [`BEGIN IMMEDIATE`](https://www.sqlite.org/lang_transaction.html). With the `database/sql`, driver this can be achieved by using [`_txlock=immediate`](https://pkg.go.dev/github.com/ncruces/go-sqlite3/driver) (making write transactions the default), then using [read-only transactions](https://pkg.go.dev/database/sql#TxOptions) for the ones that don't write (you'll get [`SQLITE_READONLY`](https://www.sqlite.org/rescode.html#readonly) if you try).

To avoid `SQLITE_LOCKED`, you need avoid dropping tables/indices _from the same connection_ and _in the middle of_ stepping queries on those tables/indices. This is a hard mistake to make with `database/sql`, so I wouldn't worry about it.

PS: if you get lots of timeouts and busy errors, do remember to `Close` your `sql.Rows`, `sql.Stmt`, and `sql.Conn`, and either `Commit` or `Rollback` your `sql.Tx`. No amount of waiting or retrying will fix these mistakes.

**Reply by NyaaaWhatsUpDoc** [reply]
This is very informative thank you! I tried removing our `retryOnBusy()` wrapping code now, and with the busy-timeout all still seems to be working as expected. We were aware of setting the busy timeout previously, but I'm not entirely certain the modernc.org/sqlite implementation behaves *entirely* as expected. (or maybe we were just using it incorrectly, i'm not sure, i don't want to point all blame at them)

**Comment by ncruces** [comment]
### ExecContext is weird

> we hide ExecerContext as it follows a weird logic path in sql.DB, being only a short-path for stmts with no args.

`Exec` can execute multiple statements (e.g. `PRAGMAS` or DDL/DML) in one go. Having no parameters and no results makes this saner, and reduces subtle bugs.

Everything else (including `Query`) goes through `Prepare` , which does the opposite and errors out if there are multiple statements.

This is my attempt at enforcing some correctness properties.

**Reply by NyaaaWhatsUpDoc** [reply]
useful to know, thank you!

**Comment by ncruces** [comment]
### Error translation

In general I'd advising going with idiomatic Go and using this pattern for error translation:
https://github.com/ncruces/go-sqlite3/blob/main/gormlite/error_translator.go

Obviously, that's just a matter of style and you may prefer to keep it how you did it. This is just to say that this driver tries to, as much as possible, respect Go idioms in error handling.

**Reply by NyaaaWhatsUpDoc** [reply]
Ah interesting! This is a much nicer way of doing things, and more idiomatic I'll agree, though in ours since it's not a publicly exported library I wasn't as worried about that and instead was striving for performance, however small the difference. (since `errors.Is()` relies on reflection)

**Comment by NyaaaWhatsUpDoc** [comment]
Update on the situation: memory leak resolved. It was me adding `QueryContext()` and `ExecContext()` support to our `SQLiteConn{}` wrapper causing it. The statements being prepared must not have ended up closed for whatever reason (though i did hook rows.Close() to do this, i was probably missing something somewhere). So entirely a "me" problem :sweat_smile: 

All (bar two things) seems to be working as expected now, with less code than we had previously for the modernc.org/sqlite implementation, since the busy timeout handler can now successfully be relied on!

The only 2 things remaining are a particularly slow running query (again this could be an issue of our own making, no fingers being pointed here), which we don't experience under modernc.org/sqlite, and our tests seem to fail using `"addr=$randomUUID?mode=memory&cache=shared"` connection string which previously we never had issues with.

**Reply by ncruces** [reply]
#### In memory (test?) databases

This driver does not support shared cache mode, as it is [discouraged](https://www.sqlite.org/sharedcache.html). There's still one remaining use for shared cache though, which is [in-memory databases that you want to share between connections](https://www.sqlite.org/inmemorydb.html).

The solution though, in recent versions of SQLite, is to [use the memdb VFS instead](https://sqlite.org/forum/forumpost/0359b21d172bd965), which I also [support](https://pkg.go.dev/github.com/ncruces/go-sqlite3/vfs/memdb).

So basically, instead of `mode=memory&cache=shared` you need `vfs=memdb`. For the database to be shared between connections, the filename needs to [start with a `/`](https://sqlite.org/src/artifact/16679def11?&ln=36-48), and if you want a clean DB at the start of the test, you should probably delete it first:
https://github.com/ncruces/go-sqlite3/blob/2c30bc996aa778a996a877b431736819f81f6e01/tests/parallel/parallel_test.go#L131-L135

PS: don't forget to `import _ "github.com/ncruces/go-sqlite3/vfs/memdb"`

**Reply by ncruces** [reply]
#### Slow queries

A slow query could easily be "our fault."

Our compiler is different ("Wasm+wazero" vs. "c2go+go") so CPU bottlenecks could differ.
Our storage engine is different (we reimplement the OS layer) so IO bottlenecks could differ.

One way we may be different (though I'm not sure, because I'm unsure of modernc internals) is if the query does large sorts.

SQLite can do sorts in parallel, if [configured to do so](https://www.sqlite.org/compile.html#max_worker_threads). We don't support this, as Wasm is single threaded (and I don't plan to change this even with threads support: single threading is a choice here).

**Reply by daenney** [reply]
I don't believe we currently set/configure worker threads in GoToSocial for SQLite, so I suspect that at least is not that.

**Reply by ncruces** [reply]
If you can capture a `pprof` of the running query, I can maybe look into it.
Although, it'll probably be a not very instructive "running native code" which is SQLite, basically.

**Reply by NyaaaWhatsUpDoc** [reply]
why the choice for single threading, even if wasm adds thread support, out of interest? (not trying to push your answer on this, just curious!).

later i'll try capture a CPU profile via pprof, take a look and share the results. again thank you for all the input here!

**Reply by ncruces** [reply]
By assuming C/Wasm is single threaded, we avoid a lot of complexity.

SQLite can be compiled without threads, without mutexes; malloc will also run without mutexes.
Each connection is sandboxed in its own address space, and shares nothing with other connections.

Basically, each SQLite instance acts like a single-threaded process isolated from the world, and needs Go to act as an OS for it to interact with outside world.

Every time I tried to change this (for shared cache, for shared memory), I got into nothing but pain.

**Comment by ncruces** [comment]
So, question: do you see this getting some traction? Any blockers remaining?

**Reply by NyaaaWhatsUpDoc** [reply]
so I've been running a build of gotosocial with your sqlite3 library for the past 4-5 days, and all has been working fine so far, bar one query (mentioned above). from my own experimenting and switching back to our old sqlite implementation, i think this is less a result of the sqlite3 implementation and more just a case of me having a 13GB database, with 6331354 rows (just checked!) in the table in question, lots of concurrent WAL connections and a tiny VPS. the main blocker right now is getting our test suite working using vfs=memdb, and updating our existing sqlite implementation to use vfs=memdb instead of the old shared memory.

once that's sorted, the plan is to implement this behind build tags in our repository so users can test it if they wish. there's a draft PR open for this but i'm still wrangling with the test failures at the moment (from switching modernc.org/sqlite to vfs=memdb mainly).

**Reply by NyaaaWhatsUpDoc** [reply]
so far it's largely @daenney and myself among the core devs really pushing for this, as we both believe embedding a wasm build is the better long-term solution than the custom transpile that is modernc.org/sqlite, and so far this library has been easier to work with / understand should we have to make any upstream PRs :)

**Reply by ncruces** [reply]
I haven't looked, but `vfs=memdb` should work with _modernc_ as well (making any transition less painful).

**Reply by daenney** [reply]
It's mostly @NyaaaWhatsUpDoc doing the actual work, I just nerd sniped her into doing it :sweat_smile:.

But yeah, the reason I started looking is because I think long term this will be more reliable and maintainable than the transpile. And we've recently had some shenanigans with the transpile, so it seemed worth exploring a potential alternative.

**Reply by ncruces** [reply]
Yeah, hopefully, long term this is only a simple recompile after every new SQLite version.

In fact, no SQLite update has created a problem so far. The only glitch I had was with a clang/wasi-sdk update that started creating named modules rather than anonymous, but I could've stayed with the older compiler.

I'd say wazero is the riskiest component (rather than me, my work is done, and easily replaceable). But wazero seems to have sustainable maintenance support, both corporate and community.

**Reply by NyaaaWhatsUpDoc** [reply]
> I haven't looked, but `vfs=memdb` should work with _modernc_ as well (making any transition less painful).

it does work, though not exactly as things currently work (since we're currently relying on shared memory mode for tests). there's lots of concurrency and now running into issues of test queries failing with "no such table" etc, the joys of concurrency ... :smiling_face_with_tear:

it's definitely the best route to go down just lots of test wrangling ahead before https://github.com/superseriousbusiness/gotosocial/pull/2863 is ready to go

(just to be clear this is with updating modernc to use vfs=memdb, after that i can i can start running the test suites with the wasm build)

**Reply by ncruces** [reply]
I think I'm finally ready to go with encryption, so I'll make a release shortly.
If you have anything else you'd like me to consider, give me a ping.

**Reply by ncruces** [reply]
Fingers crossed!
This side, I'll make a new release when [SQLite 3.46](https://sqlite.org/draft/releaselog/3_46_0.html) ships.

**Reply by daenney** [reply]
It seems the one issue we still have remaining is the public timeline query timing out when using the WASM port but it working just fine with the modernc transpile. Otherwise everything works, tests pass etc.

It's behind a build tag right now so we can safely merge this without that affecting anyone. I intend to run the WASM build myself. Maybe that's also going to help suss out what's going on with that timeline query, since then it's not just @NyaaaWhatsUpDoc's instance.

**Reply by daenney** [reply]
If we can fix that query issue then we could realistically also provide public builds with the WASM port, alongside the regular ones for now. Then people can try it out if they want to and we might get some better coverage on how this is performing overall.

If that all works out well then at some point in the future we could switch to the WASM build being our default.

**Reply by NyaaaWhatsUpDoc** [reply]
i just tried a build of WASM sqlite with the additional flag "SQLITE_ENABLE_STAT4" and it fixes the slow query issue :grin: . which i tried as i was comparing the build flags that modernc uses compared to here

though given the docs on it, i do understand why you may not want this enabled as a default.

**Reply by ncruces** [reply]
Oh that makes sense! `SQLITE_ENABLE_STAT4` is [enabled](https://gitlab.com/cznic/sqlite/-/blob/master/lib/sqlite_linux_amd64.go?ref_type=heads#L1) in modernc.

I'll review these compile options (modernc and mattn) and tweak the binary after SQLite 3.46. I've already tested the beta, and will test any flag changes, so there should be no surprises.

That said. You can _always_ use your own custom build, if you have specific needs. My intent is that this is _explicitly_ supported. But I also want to provide a good default experience.

**Reply by NyaaaWhatsUpDoc** [reply]
Exactly that was my thoughts. Tweaking the opts to our needs wouldn't be too difficult as a fork to keep up to date (if that's even the best way of doing it), there may be easier ways. The only part of `SQLITE_ENABLE_STAT4` I wasn't sure if you'd want to make as a default opt was the fact they mention that "the downside of this compile-time option is that it violates the [query planner stability guarantee](https://www.sqlite.org/queryplanner-ng.html#qpstab) making it more difficult to ensure consistent performance in mass-produced applications.".

**Reply by daenney** [reply]
I think by default query stability guarantee isn't enabled even in SQLite, you have to set `SQLITE_ENABLE_QPSG`. So unless this build already does that, enabling STAT4 shouldn't have any negative consequences.

**Reply by ncruces** [reply]
The options modernc includes and I don't (mattn doesn't either, for the most part) are:
- [x] `SQLITE_ENABLE_STAT4`
- [x] `SQLITE_ENABLE_COLUMN_METADATA`
- [ ] `SQLITE_ENABLE_DBSTAT_VTAB`
- [ ] `SQLITE_ENABLE_MEMORY_MANAGEMENT`
- [ ] `SQLITE_ENABLE_OFFSET_SQL_FUNC`
- [ ] `SQLITE_ENABLE_SNAPSHOT`
- [ ] `SQLITE_ENABLE_SESSION`
- [ ] `SQLITE_ENABLE_PREUPDATE_HOOK`
- [ ] `SQLITE_ENABLE_RBU`

For now I'm inclined to include `SQLITE_ENABLE_STAT4`. The others I may include upon request. Most require APIs for wrapping, so if you find this via search, open an issue. Some are not useful with a Wasm build. And usefulness will be weighed against Wasm code size increase.

**Reply by ncruces** [reply]
> Exactly that was my thoughts. Tweaking the opts to our needs wouldn't be too difficult as a fork to keep up to date (if that's even the best way of doing it), there may be easier ways.

You don't need to fork. You need to create your own version of package [`embed`](https://github.com/ncruces/go-sqlite3/tree/main/embed), and maybe pin a version in `go.mod` for stability. Or ask me to provide (e.g.) `embed/stat4`, and support it.

> I think by default query stability guarantee isn't enabled even in SQLite, you have to set `SQLITE_ENABLE_QPSG`. So unless this build already does that, enabling STAT4 shouldn't have any negative consequences.

I don't. And you can `DBCONFIG_ENABLE_QPSG` if you want query stability. The only question is how much `SQLITE_ENABLE_STAT4` bloats the Wasm (I don't expect very much). But every little feature does, a bit.

And the only big downside of bloating the Wasm, is that opening the first DB connection becomes slower, as the wazero compiler is a little slow.

**Comment by ncruces** [comment]
So, this got merged in https://github.com/superseriousbusiness/gotosocial/pull/2863!

If you're ever serious about [releasing](https://github.com/superseriousbusiness/gotosocial/releases) a FreeBSD binary with this enabled, something needs to be done about #85.

Otherwise, what you have should be good for both Linux and macOS.

**Reply by daenney** [reply]
Right, GtS is packaged on the BSDs :sweat_smile:. Thanks for the shout, we need to keep that in mind.
